exercises/src/main/scala/fpinscala/laziness/Stream.scala

```diff
@@ -17,20 +27,95 @@ trait Stream[+A] {
     case Empty => None
     case Cons(h, t) => if (f(h())) Some(h()) else t().find(f)
   }
-  def take(n: Int): Stream[A] = ???
 
-  def drop(n: Int): Stream[A] = ???
+  //5.2
+  def take(n: Int): Stream[A] = this match {
+    case _ if n == 0 => empty
+    case Cons(h, _) if n == 1 => cons(h(), empty)
+    case Cons(h, t) => cons(h(), t().take(n - 1))
+  }
+
+  //5.2
+  def drop(n: Int): Stream[A] = this match {
+    case Cons(_, t) if n >= 1 => t().drop(n - 1)
+    case _ => this
+  }
+
+  //5.3
+  def takeWhile(p: A => Boolean): Stream[A] = this match {
+    case Cons(h, t) if p(h()) => cons(h(), t().takeWhile(p))
+    case _ => empty
+  }
 
-  def takeWhile(p: A => Boolean): Stream[A] = ???
+  //5.4
+  def forAll(p: A => Boolean): Boolean = foldRight(true)((a, b) => p(a) && b )
 
-  def forAll(p: A => Boolean): Boolean = ???
+  //5.5 takeWhile with foldRight
+  def takeWhile2(p: A => Boolean): Stream[A] = foldRight(Empty: Stream[A])((a, b) => if (p(a)) cons(a, b) else Empty)
 
-  def headOption: Option[A] = ???
+  //5.6 headOption with foldRight
+  def headOption: Option[A] = foldRight(None: Option[A])((a, _) => Some(a))
 
   // 5.7 map, filter, append, flatmap using foldRight. Part of the exercise is
   // writing your own function signatures.
+  def map[B](f: A => B): Stream[B] = foldRight(Empty: Stream[B])((a, b) => cons(f(a), b))
+
+  def filter(f: A => Boolean): Stream[A] = foldRight(Empty: Stream[A])((a, b) => if (f(a)) cons(a, b) else b)
+
+  def append[B >: A](x: => Stream[B]): Stream[B] = foldRight(x)((a, b) => cons(a, b))
+
+  def flatMap[B](f: A => Stream[B]): Stream[B] = foldRight(Empty: Stream[B])((a, b) => f(a) append b)
+
+  //5.13 map, take, takeWhile, zipWith, zipAll using unfold
+  def map2[B](f: A => B): Stream[B] = unfold(this) {
+    case Cons(h, t) => Some(f(h()), t())
+    case Empty => None
+  }
+
+  def take2(n: Int): Stream[A] = unfold((this, n)) {
+    case (Cons(h, t), x) if x > 0 => Some(h(), (t(), x - 1))
+    case _ => None
+  }
```

Comment:
If you go back to the question I wrongly asked about the first implementation of `take`, this one behave differently as for as being strict vs lazy on the tail of the last element.